### DISCUSSION TITLE
Dealing with duplicates on a many-to-many relationship

### DISCUSSION THREAD

**Discussion by jtpavlock** [sqlalchemy/sqlalchemy#6731] [Usage Questions] [answered]
I think a code example will speak a thousand words here:

```python
import sqlalchemy
from typing import List
from sqlalchemy import Column, Integer, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, sessionmaker
from sqlalchemy.schema import ForeignKey, Table
from sqlalchemy.ext.associationproxy import association_proxy

Session = sessionmaker()
Base = declarative_base()


class _Genre(Base):
    """A track can have multiple genres."""

    __tablename__ = "genre"

    name: str = Column(String, nullable=False, primary_key=True)

    def __init__(self, name: str):
        self.name = name


track_genre = Table(
    "track_genre",
    Base.metadata,
    Column("genre", String, ForeignKey("genre.name")),
    Column("track_id", Integer, ForeignKey("track.id")),
)
__table_args__ = ()


class Track(Base):
    __tablename__ = "track"

    id: int = Column(Integer, primary_key=True)

    _genres: List[_Genre] = relationship(
        "_Genre",
        secondary=track_genre,
        collection_class=list,
        cascade_backrefs=False,
    )
    genre: List[str] = association_proxy("_genres", "name")

    def __init__(self, id):
        self.id = id


engine = sqlalchemy.create_engine("sqlite:///:memory:")
Session.configure(bind=engine)
Base.metadata.create_all(engine)
session = Session()

session.add(Track(id=1))
session.add(Track(id=2))

tracks = session.query(Track).all()
for track in tracks:
    track.genre = ["pop"]

session.commit()
```

The above results in an error `sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: genre.name` and I am wondering what is the best way to do what I am attempting? I tried adding `sqlite_on_conflict_primray_key="IGNORE"` to `genre.name` column, which solves the error, but then produces warnings
```python
/home/jacob/src/Moe/sqlalchemy_mre.py:70: SAWarning: New instance <_Genre at 0x7f7892907670> with identity key (<class '__main__._Genre'>, ('pop',), None) conflicts with persistent instance <_Genre at 0x7f78928a67f0>
  session.commit()
/home/jacob/src/Moe/sqlalchemy_mre.py:70: SAWarning: Identity map already had an identity for (<class '__main__._Genre'>, ('pop',), None), replacing it with newly flushed object.   Are there load operations occurring inside of an event handler within the flush?
  session.commit()
```

I could just suppress the warnings if they are harmless, but I am wondering if I am doing something wrong or if there is a better way.

**Comment by zzzeek** [comment, marked as answer]
if you want to work from a strictly SQL perspective, then the INSERT into the "genre" table that's ignored is not a problem.  but if you want to work from the ORM, it needs there to be just one instance of an object for a particular primary key.

the warnings you see are because you've created two separate Genre() objects with the same primary key in one session.  if you were to ``session.expunge()`` one of them first before persisting the next one, then you wouldn't see this warning, but that's still awkward.

the canonical way to handle this is to use the same `_Genre` object per session if you are referring to that same primary key more than once.  you can get any object from the Session of a particular primary key using ``session.get``, which will not use any SQL if the object is already there.

it's possible to customize what the association proxy does when you append to it, using the creator argument discussed at https://docs.sqlalchemy.org/en/14/orm/extensions/associationproxy.html?highlight=association%20proxy#creation-of-new-values .    Historically we would tell people to use a session lookup inside the creator, using recipes like [unique object](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) which you should check out.   But the downside is that to create a "magical" pattern like the one you have above, you would have to use a globally available session of some kind like a thread local "scoped session".   that might be all you want to do here though.

another way to do it, since you have a really simple object here, is to intercept that object before it gets added to the session and produces problems, but in a hook where you can get at the relevant session.    that can be done here using a [@validates](https://docs.sqlalchemy.org/en/14/orm/mapped_attributes.html?highlight=validates#simple-validators) method like this, this is a little more of a novel method but for something simple like this, it's fine (if things get more complicated, maybe not):

```py
class Track(Base):
    __tablename__ = "track"

    id: int = Column(Integer, primary_key=True)

    _genres: List[_Genre] = relationship(
        "_Genre",
        secondary=track_genre,
        collection_class=list,
        cascade_backrefs=False,
    )
    genre: List[str] = association_proxy("_genres", "name")

    @validates("_genres")
    def _append_genre(self, key, genre):
        sess = object_session(self)
        persistent_genre = sess.get(_Genre, genre.name)
        if persistent_genre:
            # replace the genre that was given with one that is persistent
            # already.  this assumes the given genre is simple and doesnt
            # have other important state you wouldn't want to throw away,
            # since we are basically throwing it away here
            return persistent_genre
        else:
            # otherwise persist the genre
            sess.add(genre)

            # if autoflush isn't used, make sure this genre object is flushed.
            sess.flush([genre])

            return genre

    def __init__(self, id):
        self.id = id
```

**Reply by jtpavlock** [reply]
Thank you so much for the quick and extremely helpful answer!

I opted to use your given example with the `validates` method and just add the following to it
```python
if not sess:
    return genre
```
